### ISSUE TITLE
Python 3.7

### ISSUE THREAD

**Issue by mjholtkamp** [iusrepo/wishlist#200]
### What new package do you want?

python 3.7

### Why?

python 3.7 has been released recently and it provides several new features and performance improvements. We were happy with the python3.5 and python3.6 packages from ius, so we would like the 3.7 version as well.

### Testing

I agree to test the new package to ensure that it works as expected.  Once I am
satisfied with the results of my testing I will comment on this issue with the
word "STABLE" to get it promoted to the stable repos.

**Comment by carlwgeorge** [comment]
Thanks @mjholtkamp for submitting this request.  This certainly seems like a great addition to IUS.  I'll get to work on this as soon as I'm able.

**Comment by moggers87** [comment]
Just as a heads up, my understanding is that Python 3.7 will require a newer version of OpenSSL than RHEL/CentOS 6 provides

From the SSL module docs:

> Changed in version 3.6: OpenSSL 0.9.8, 1.0.0 and 1.0.1 are deprecated and no longer supported. In the future the ssl module will require at least OpenSSL 1.0.2 or 1.1.0.

And from a CentOS 6 VM:

```
# rpm -q openssl                                                                                                   
openssl-1.0.1e-57.el6.i686
```

**Comment by moggers87** [comment]
Sure, making a parallel installable package of a newer version OpenSSL for RHEL 6 is possible but it's more work than just creating the Python 3.7 package ☺

**Comment by mjholtkamp** [comment]
Sad to hear, but I understand the reasoning. Hopefully this will give us more push to move away from an EL6-based distro to a newer version.

**Comment by fatalcoder** [comment]
@carlwgeorge  any plans when python37u will be available for RHEL7?

**Comment by FlipperPA** [comment]
@carlwgeorge I'm wondering if there's anything we can do to help with getting the Python 3.7 (and soon, 3.8) suite of packages (python37u-test, python37u-mod_wsgi, etc) included. Is there a walkthrough somewhere where we could take the `python36u` repo, fork it to `python37u` and test building the packages so it doesn't all fall on you?

**Comment by tomamoto** [comment]
For what it's worth, CentOS7's (and presumably RHEL7 as well) base repo now includes OpenSSL 1.0.2.  As of this posting (from `yum info openssl`):
Installed Packages
Name        : openssl
Arch        : x86_64
Epoch       : 1
Version     : 1.0.2k
Release     : 12.el7
Size        : 814 k
Repo        : installed
From repo   : base

**Comment by chrisoro** [comment]
Please keep in mind that depending on how long this will take, just using python 3.8 might be better.
It is expected to be release in October (https://www.python.org/dev/peps/pep-0569/).

Just dropping since here since there was no update for more than half a year and it seems unclear if 37 will ever happen here.

**Comment by FlipperPA** [comment]
I greatly appreciate the efforts of everyone at IUS community, but does anyone know of another similar service with more current versions of yum-installable Python packages?

**Comment by carlwgeorge** [comment]
Sorry for the delays here, I've been working through some [necessary infrastructure work](https://github.com/iuscommunity/wishlist/issues/219#issuecomment-481516179).

Signs indicate that [Red Hat is planning to add a Python 3 stack to RHEL7](https://bugzilla.redhat.com/show_bug.cgi?id=1639030).  This will likely be a tech preview to assist users in migrating to RHEL 8.  I don't have any inside knowledge here as I don't work for Red Hat, but based on comments in multiple bugzilla issues and conversations in the Fedora Python SIG, I don't think this is very far out.  It will likely be Python 3.6, but there is an outside chance they could choose to deliver Python 3.7 instead.  I'm going to hold off on the work to create a python37 package until that is determined.

This will completely change the landscape of Python 3 in RHEL7, which currently includes multiple python3X packages from IUS (python34u, python35u, python36u) and EPEL (python34, python36).  I joined the Fedora Python SIG with the intention of reducing duplicate work and possible having the EPEL packages obsolete their IUS counterparts.  That may still be worthwhile, but to be completely honest a recent Python 3 in the base distribution will render earlier versions of Python mostly pointless.

So to summarize, if RHEL7 gets Python 3.7, adding python37 to IUS doesn't makes sense.  If RHEL7 ships Python 3.6, then a python37 package would make sense, either in IUS or EPEL (parallel installable packages meet EPEL guidelines).  If the Fedora Python SIG is on board with the idea, I would prefer EPEL as I would have more co-maintainers.  In fact, python37 already exists as a alternate stack in Fedora 28.

**Comment by FlipperPA** [comment]
Thanks for the update, @carlwgeorge, and your efforts... they are very much appreciated. The amount of time and headaches you've saved me (and a whole load of non-profits I help!) is substantial.

It would be great if you could get some support, but I must admit, I am wary of RedHat keeping current. Python 3.8 is due in October, so that will be a good test. I typically install both the EPEL and IUS repos, so as long as a close-to-current version appears in either, I'll be a happy camper.

**Comment by Chaz6** [comment]
FWIW epel now ships Python 3.6 for CentOS 7.

**Comment by tlusk** [comment]
Here's a patch against Fedora's 3.7.3-3 RPM to port it to EL6/7.  The patch makes some opinionated changes such as building with the RHEL developer toolset and some things for OpenSSL on EL6, but other than those it should be a good starting point for people wanting Python 3.7 on EL6/7

https://gist.github.com/tlusk/cd2a395f0da58cd5301e60a80c4ac89b

**Comment by FlipperPA** [comment]
@Chaz6 Do you happen to know if that version that ships with EPEL includes `--enable-optimizations`?

**Comment by lauriliba** [comment]
@FlipperPA Yes it does. You can verify with the sysconfig module
```
❯ docker run -it --rm centos:7 /bin/bash
[root@b3e03b6e7209 /]# yum install -y -q epel-release && yum install -y -q python36
warning: /var/cache/yum/x86_64/7/extras/packages/epel-release-7-11.noarch.rpm: Header V3 RSA/SHA256 Signature, key ID f4a80eb5: NOKEY
Public key for epel-release-7-11.noarch.rpm is not installed
Importing GPG key 0xF4A80EB5:
 Userid     : "CentOS-7 Key (CentOS 7 Official Signing Key) <security@centos.org>"
 Fingerprint: 6341 ab27 53d7 8a78 a7c2 7bb1 24c6 a8a7 f4a8 0eb5
 Package    : centos-release-7-6.1810.2.el7.centos.x86_64 (@CentOS)
 From       : /etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7
warning: /var/cache/yum/x86_64/7/epel/packages/python36-3.6.8-1.el7.x86_64.rpm: Header V3 RSA/SHA256 Signature, key ID 352c64e5: NOKEY
Public key for python36-3.6.8-1.el7.x86_64.rpm is not installed
Importing GPG key 0x352C64E5:
 Userid     : "Fedora EPEL (7) <epel@fedoraproject.org>"
 Fingerprint: 91e9 7d7c 4a5e 96f1 7f3e 888f 6a2f aea2 352c 64e5
 Package    : epel-release-7-11.noarch (@extras)
 From       : /etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-7
[root@b3e03b6e7209 /]# python3
Python 3.6.8 (default, Apr 25 2019, 21:02:35) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-36)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sysconfig
>>> print(sysconfig.get_config_var('CONFIG_ARGS'))
'--build=x86_64-redhat-linux-gnu' '--host=x86_64-redhat-linux-gnu' '--program-prefix=' '--disable-dependency-tracking' '--prefix=/usr' '--exec-prefix=/usr' '--bindir=/usr/bin' '--sbindir=/usr/sbin' '--sysconfdir=/etc' '--datadir=/usr/share' '--includedir=/usr/include' '--libdir=/usr/lib64' '--libexecdir=/usr/libexec' '--localstatedir=/var' '--sharedstatedir=/var/lib' '--mandir=/usr/share/man' '--infodir=/usr/share/info' '--enable-ipv6' '--enable-shared' '--with-computed-gotos=yes' '--with-dbmliborder=gdbm:ndbm:bdb' '--with-system-expat' '--with-system-ffi' '--enable-loadable-sqlite-extensions' '--with-dtrace' '--with-valgrind' '--without-ensurepip' '--enable-optimizations' 'build_alias=x86_64-redhat-linux-gnu' 'host_alias=x86_64-redhat-linux-gnu' 'CFLAGS=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv  ' 'LDFLAGS=-Wl,-z,relro  -g  ' 'CPPFLAGS= ' 'PKG_CONFIG_PATH=:/usr/lib64/pkgconfig:/usr/share/pkgconfig'
>>> 
```

**Comment by FlipperPA** [comment]
@lauriliba Thank you! Today I learned `sysconfig.get_config_var('CONFIG_ARGS')`. :)

**Comment by carlwgeorge** [comment]
RHEL 7.7 was released with python3 at version 3.6.  [I've got some work to do](https://github.com/iusrepo/packaging/issues/7), but once that's complete I'm going to work with the Fedora Python SIG to add python37 to EPEL.

**Comment by xkrt** [comment]
@carlwgeorge is there some ticket "python37 in EPEL" to watch?

**Comment by carlwgeorge** [comment]
@xkrt Not yet.  I'm waiting for CentOS 7.7 to be released to start that conversation with the Fedora Python SIG, as they are pretty busy with preparing to retire EPEL7's python36 (obsoleted by RHEL 7.7's python3).
